src/AzureExtension/DataManager/CacheManager.cs

```diff
@@ -0,0 +1,333 @@
+            {
+                _log.Information("Cancelling update.");
+                _cancelSource.Cancel();
+            }
+        }
+    }
+
+    public async Task Refresh()
+    {
+        lock (_stateLock)
+        {
+            if (_pendingRefresh)
+            {
+                _log.Debug("Refresh already pending, ignoring refresh request.");
+                return;
+            }
+
+            _pendingRefresh = true;
+            _clearNextDataUpdate = true;
+        }
+
+        CancelUpdateInProgress();
+        await Update(TimeSpan.MinValue);
+    }
+
+    public IEnumerable<IRepository> GetRepositories()
+    {
+        var devHomeRepositories = new List<IRepository>();
+        var repositories = DataManager.GetRepositories();
+        foreach (var repository in repositories)
+        {
+            // Convert data model repositories, which have datastore connections
+            // and table lookups as part of the data model, to a static snapshot
+            // of the repository data. This will be sent to DevHome, so this is
+            // needs to be detached data from our data store.
+            devHomeRepositories.Add(new DevHomeRepository(repository));
+        }
+
+        return devHomeRepositories;
+    }
+
+    private async Task PeriodicUpdate()
+    {
+        // Only update per the update interval.
+        // This is intended to be dynamic in the future.
+        if (DateTime.Now - LastUpdateTime < UpdateInterval)
+        {
+            return;
+        }
+
+        try
+        {
+            _log.Debug("Starting account data update.");
+            await Update(DefaultAccountUpdateFrequency);
+        }
+        catch (Exception ex)
+        {
+            Log.Error(ex, "Update failed unexpectedly.");
+        }
+
+        LastUpdateTime = DateTime.Now;
+        return;
+    }
+
+    private async Task Update(TimeSpan? olderThan)
+    {
+        var options = new RequestOptions();
+
+        lock (_stateLock)
+        {
+            if (UpdateInProgress)
+            {
+                _log.Information("Update is in progress, ignoring request.");
+                return;
+            }
+
+            UpdateInProgress = true;
+            _cancelSource = new CancellationTokenSource();
+            options.CancellationToken = _cancelSource.Token;
+            options.Refresh = _clearNextDataUpdate;
+        }
+
+        _log.Debug("Starting account data update.");
+        SendUpdateEvent(this, CacheManagerUpdateKind.Started);
+        await DataManager.UpdateDataForAccountsAsync(olderThan ?? DefaultAccountUpdateFrequency, options, Id);
+    }
+
+    private void HandleDataManagerUpdate(object? source, DataManagerUpdateEventArgs e)
+    {
+        if (e.Requestor == Id)
+        {
+            Log.Debug("DataManager update received");
+            switch (e.Kind)
+            {
+                case DataManagerUpdateKind.Cache:
+                    lock (_stateLock)
```

Comment:
NIT: These cases are long.  Can this be converted to if/else instead?  I find long blocks of code easier on if/else.

Reply:
I disagree. This is a switch based on an enum, which means it will not compile if cases are missing (or there is a default) or one of the cases isn't valid. This ensures if the enum changes later this code will force the case to be added. So for maintainability this is a better choice. Personally I find the switch statement cleaner than nested if/elses and the logic easier to follow since C# does not allow fallthrough, especially for enums.